### PR TITLE
Added tests for text featurizer options (Part2).

### LINKED ISSUE #2967: More tests for TextFeaturizer.
TextFeaturizer is a complex transform and it takes more options as input then any other component in ML.NET. Using those options, the TextFeaturizer build up the underlying components.

Assuming that underlying component are well tested, there is a need to test the options for TextFeaturizer so that we are sure that underlying components are created as intended and results produced are as expected.

### PR DESCRIPTION
**Pull request by zeahmed**
This PR finally fixes #2967. Test created in this PR are for the following parameters in options class
* WordNgramExtractor
* CharNgramExtractor
* Numeric Feature Normalizer (L1, L2, etc).

The intend here is to test that TextFeaturizer is instantiated for every parameter in the options class. Here, we are not testing the internal components of TextFeaturizer.

### CODE REVIEW

File: test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs [hunk 1]
```diff
@@ -51,6 +52,139 @@ public void TextFeaturizerWithPredefinedStopWordRemoverTest()
...
+                Norm = TextFeaturizingEstimator.NormFunction.None,
+                OutputTokensColumnName = "OutputTokens"
+            };
+            var pipeline = ML.Transforms.Text.FeaturizeText("Features", options, "A");
+            var model = pipeline.Fit(dataView);
+            var engine = model.CreatePredictionEngine<TestClass, TestClass>(ML);
+
+            var prediction = engine.Predict(data[0]);
+            Assert.Equal("this is some text in english", string.Join(" ", prediction.OutputTokens));
+            Assert.Equal(1.0f, prediction.Features[0]);
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs]: >Assert.Equal(1.0f, prediction.Features[0]) [](start = 12, length = 42)

Doesn't `Assert` has  option to compare to arrays or enumerables?

File: test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs [hunk 2]
```diff
@@ -51,6 +52,139 @@ public void TextFeaturizerWithPredefinedStopWordRemoverTest()
...
+            Assert.Equal(1.0f, prediction.Features[6]);
+            Assert.Equal(1.0f, prediction.Features[7]);
+            Assert.Equal(1.0f, prediction.Features[8]);
+            Assert.Equal(0.0f, prediction.Features[9]);
+            Assert.Equal(0.0f, prediction.Features[10]);
+            Assert.Equal(0.0f, prediction.Features[11]);
+
+            prediction = engine.Predict(data[1]);
+            Assert.Equal("xyz", string.Join(" ", prediction.OutputTokens));
+            Assert.Equal(1.0f, prediction.Features[0]);
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs]: >Assert.Equal(1.0f, prediction.Features[0]); [](start = 12, length = 43)

So i expect ngrams to be `a`,`b`,`c`,`e`,`f`,`g`,`x`,`y`,`z`,` ` and end of string and empty string, not sure honestly,  I guess. total 12 ngrams.
feature 0 and feature 8 is this end of string and empty string, right?

Comment (zeahmed) [reply test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs]: Index 0 is start marker and index 8 is end marker. Then there are total 10 characters including space.

---
In reply to: [267486364](https://github.com/dotnet/machinelearning/pull/3036#discussion_r267486364) [](ancestors = 267486364)

File: test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs [hunk 3]
```diff
@@ -51,6 +52,95 @@ public void TextFeaturizerWithPredefinedStopWordRemoverTest()
...
+                CharFeatureExtractor = null,
+                Norm = TextFeaturizingEstimator.NormFunction.None,
+                OutputTokensColumnName = "OutputTokens"
+            };
+            var pipeline = ML.Transforms.Text.FeaturizeText("Features", options, "A");
+            var model = pipeline.Fit(dataView);
+            var engine = model.CreatePredictionEngine<TestClass, TestClass>(ML);
+
+            var prediction = engine.Predict(data[0]);
+            Assert.Equal("this is some text in english", string.Join(" ", prediction.OutputTokens));
```
Comment (sfilipi) [inline test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs]: >"this is some text in english" [](start = 25, length = 30)

nit, but for maintainability i'd create a var for this.

File: test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs [hunk 4]
```diff
@@ -51,6 +52,95 @@ public void TextFeaturizerWithPredefinedStopWordRemoverTest()
...
+            var expected = new float[] { 1.0f, 1.0f, 1.0f, 1.0f, 1.0f, 1.0f, 0.0f, 0.0f };
```
Comment (sfilipi) [inline test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs]: >0.0f, 0.0f [](start = 77, length = 10)

Curious: why do we have those two at the end?

File: test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs [hunk 5]
```diff
@@ -51,6 +52,95 @@ public void TextFeaturizerWithPredefinedStopWordRemoverTest()
...
+                Norm = TextFeaturizingEstimator.NormFunction.None,
+                OutputTokensColumnName = "OutputTokens"
+            };
+            var pipeline = ML.Transforms.Text.FeaturizeText("Features", options, "A");
+            var model = pipeline.Fit(dataView);
+            var engine = model.CreatePredictionEngine<TestClass, TestClass>(ML);
+
+            var prediction = engine.Predict(data[0]);
+            Assert.Equal("abc efg", string.Join(" ", prediction.OutputTokens));
+            var expected = new float[] { 1.0f, 1.0f, 1.0f, 1.0f, 1.0f, 1.0f, 1.0f, 1.0f, 1.0f, 0.0f, 0.0f, 0.0f };
```
Comment (sfilipi) [inline test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs]: >1.0f [](start = 89, length = 4)

should this be 0? or is this the end marker?

File: test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs [hunk 6]
```diff
@@ -51,6 +52,95 @@ public void TextFeaturizerWithPredefinedStopWordRemoverTest()
...
+        [Fact]
+        public void TextFeaturizerWithL2NormTest()
+        {
+            var data = new[] { new TestClass() { A = "abc xyz", OutputTokens=null},
+                               new TestClass() { A = "xyz", OutputTokens=null } };
+            var dataView = ML.Data.LoadFromEnumerable(data);
+
+            var options = new TextFeaturizingEstimator.Options()
+            {
+                CharFeatureExtractor = new WordBagEstimator.Options() { NgramLength = 1},
```
Comment (sfilipi) [inline test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs]: >CharFeatureExtractor = new WordBagEstimator.Options()  [](start = 16, length = 54)

Is this correct? it doesn't read right to initialize a CharExtractor with the options of a WordBagEstimator...

File: test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs (hunk 5 above)
Comment (zeahmed) [reply test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs]: Its the end marker.

---
In reply to: [267967436](https://github.com/dotnet/machinelearning/pull/3036#discussion_r267967436) [](ancestors = 267967436)

File: test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs (hunk 6 above)
Comment (zeahmed) [reply test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs]: Nice catch! I have added this as a part of issue in [https://github.com/dotnet/machinelearning/issues/2895](https://github.com/dotnet/machinelearning/issues/2895).

---
In reply to: [267968148](https://github.com/dotnet/machinelearning/pull/3036#discussion_r267968148) [](ancestors = 267968148)

**Pull request merged**